test/point_quest/quests/commands/attack_test.exs

```diff
@@ -0,0 +1,280 @@
+
+      params = %{quest_id: quest_id, adventurer_id: adventurer_id, attack: attack}
+
+      assert %Attack{quest_id: ^quest_id, adventurer_id: ^adventurer_id, attack: ^attack} =
+               Attack.new!(params)
+    end
+  end
+
+  describe "execute/2" do
+    setup do
+      {:ok, %{party_leader: party_leader} = quest} =
+        StartQuest.new!(%{name: "proper quest"}) |> StartQuest.execute()
+
+      {:ok, other_quest} =
+        StartQuest.new!(%{
+          name: "Steve's Shenanigans",
+          party_leaders_adventurer: %{name: "Stevey Beevey", class: :mage}
+        })
+        |> StartQuest.execute()
+
+      {:ok, %{adventurers: [adventurer | _rest]} = quest} =
+        AddAdventurer.new!(%{name: "Sir Stephen Bolton", class: :knight, quest_id: quest.id})
+        |> AddAdventurer.execute()
+
+      party_leader_actor = %PointQuest.Authentication.Actor.PartyLeader{
+        quest_id: quest.id,
+        leader_id: party_leader.id,
+        adventurer: nil
+      }
+
+      adventurer_actor = %PointQuest.Authentication.Actor.Adventurer{
+        quest_id: quest.id,
+        adventurer: adventurer
+      }
+
+      other_actor = %PointQuest.Authentication.Actor.PartyLeader{
+        quest_id: other_quest.id,
+        leader_id: other_quest.party_leader.id,
+        adventurer: other_quest.party_leader.adventurer
+      }
+
+      {:ok,
+       %{
+         quest: quest,
+         other_quest: other_quest,
+         party_leader: party_leader,
+         adventurer: adventurer,
+         party_leader_actor: party_leader_actor,
+         adventurer_actor: adventurer_actor,
+         other_actor: other_actor
+       }}
+    end
+
+    test "returns updated quest on success", %{
+      quest: %{id: quest_id},
+      adventurer: %{id: adventurer_id},
+      adventurer_actor: adventurer_actor
+    } do
+      Phoenix.PubSub.subscribe(PointQuestWeb.PubSub, quest_id)
+
+      attacked_event = %PointQuest.Quests.Event.AdventurerAttacked{
+        quest_id: quest_id,
+        adventurer_id: adventurer_id,
+        attack: 3
+      }
+
+      assert {:ok,
+              %PointQuest.Quests.Quest{
+                id: ^quest_id,
+                adventurers: [%{id: ^adventurer_id}],
+                attacks: [%PointQuest.Quests.Attack{adventurer_id: ^adventurer_id, attack: 3}]
+              },
+              ^attacked_event} =
+               Attack.new!(%{quest_id: quest_id, adventurer_id: adventurer_id, attack: 3})
+               |> Attack.execute(adventurer_actor)
+
+      assert_receive ^attacked_event, 500
+    end
+
+    test "succeeds if party leader is adventurer attacking", %{other_actor: actor} do
+      assert {:ok, _quest, _event} =
+               Attack.new!(%{
+                 quest_id: actor.quest_id,
+                 adventurer_id: actor.adventurer.id,
+                 attack: 8
+               })
+               |> Attack.execute(actor)
+    end
+
+    test "policy check fails if adventurer is not part of current quest", %{
```

Comment:
applies to all policy checks here, but they're literally too large for me to highlight them all?

These should probably be in a separate describe block specifically targeting the policy checks... but I didn't want to duplicate the setup block.

Actually as I'm typing this I now realize that this is the only setup block I was using so I can just move it up to the root level and allow both describe blocks to consume it... huh. Fixing.

Reply:
... I'm kind of amazed that github actually managed to keep this targeted to the right code block after I refactored?

Reply:
... I'm kind of amazed that github actually managed to keep this targeted to the right code block after I refactored?